File: src/main/java/gargoyle/ct/util/Defend.java
package gargoyle.ct.util;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.math.BigDecimal;
import java.text.MessageFormat;
import java.util.Objects;

@SuppressWarnings("StaticMethodOnlyUsedInOneClass")
public final class Defend {
    private Defend() {
    }

    public static <N> void equals(N actual, N expected, @NotNull String message, Object... arguments) {
        if (!Objects.equals(actual, expected)) {
            fail(message);
        }
    }

    private static void fail(@NotNull String message, Object... arguments) {
        throw new IllegalArgumentException(arguments.length == 0 ? message : MessageFormat.format(message, arguments));
    }

    public static <N extends Number> void inRange(N value, N minimum, N maximum, @NotNull String message, Object... arguments) {
        if (new BigDecimal(String.valueOf(minimum)).compareTo(new BigDecimal(String.valueOf(value))) > 0 ||
                new BigDecimal(String.valueOf(maximum)).compareTo(new BigDecimal(String.valueOf(value))) < 0) {
            fail(message);
        }
    }

    public static <N extends Comparable<N>> void inRange(@NotNull N value, N minimum, @NotNull N maximum, @NotNull String message, Object... arguments) {
        if (minimum.compareTo(value) > 0 || maximum.compareTo(value) < 0) {
            fail(message);
        }
    }

    public static void isFalse(boolean condition, @NotNull String message, Object... arguments) {
        if (condition) {
            fail(message);
        }
    }

    public static void isTrue(boolean condition, @NotNull String message, Object... arguments) {
        if (!condition) {
            fail(message);
        }
    }

    public static void notEmpty(@Nullable CharSequence value, @NotNull String message, Object... arguments) {
        if (value == null || value.length() == 0) {
            fail(message);
        }
    }

    public static void notEmptyTrimmed(@Nullable String value, @NotNull String message, Object... arguments) {
        if (value == null || value.isEmpty() || value.trim().isEmpty()) {
            fail(message);
        }
    }

    public static <T> void instanceOf(T value, Class<? extends T> type, @NotNull String message, Object... arguments) {
        if (!type.isInstance(value)) {
            fail(message, arguments);
        }
    }

    public static <T> void notNull(@Nullable T value, @NotNull String message, Object... arguments) {
        if (value == null) {
            fail(message);
        }
    }

    public static <T> void numeric(@NotNull String value, @NotNull String message, Object... arguments) {
        try {
            notNull(new BigDecimal(value), message);
        } catch (NumberFormatException e) {
            fail(message, arguments);
        }
    }
}

File: src/main/kotlin/gargoyle/ct/util/Defend.kt
package gargoyle.ct.util

import java.math.BigDecimal
import java.text.MessageFormat
import kotlin.reflect.KClass

object Defend {
    fun <N> equals(actual: N, expected: N, message: String) {
        if (actual != expected) {
            fail(message)
        }
    }

    private fun fail(message: String, vararg arguments: Any) {
        throw IllegalArgumentException(if (arguments.isEmpty()) message else MessageFormat.format(message, *arguments))
    }

    fun <N : Number?> inRange(value: N, minimum: N, maximum: N, message: String) {
        if (BigDecimal(minimum.toString()) > BigDecimal(value.toString()) ||
            BigDecimal(maximum.toString()) < BigDecimal(value.toString())
        ) {
            fail(message)
        }
    }

    fun <N : Comparable<N>?> inRange(value: N, minimum: N, maximum: N, message: String) {
        if (minimum!! > value || maximum!! < value) {
            fail(message)
        }
    }

    fun isFalse(condition: Boolean, message: String) {
        if (condition) {
            fail(message)
        }
    }

    fun isTrue(condition: Boolean, message: String) {
        if (!condition) {
            fail(message)
        }
    }

    fun notEmpty(value: CharSequence?, message: String) {
        if (value == null || value.isEmpty()) {
            fail(message)
        }
    }

    fun notEmptyTrimmed(value: String?, message: String) {
        if (value == null || value.isBlank()) {
            fail(message)
        }
    }

    fun <T : Any> instanceOf(value: T, type: KClass<out T>, message: String, vararg arguments: Any) {
        if (!type.isInstance(value)) {
            fail(message, *arguments)
        }
    }

    fun <T> notNull(value: T?, message: String) {
        if (value == null) {
            fail(message)
        }
    }

    fun <T> numeric(value: String, message: String, vararg arguments: Any) {
        try {
            notNull(BigDecimal(value), message)
        } catch (e: NumberFormatException) {
            fail(message, *arguments)
        }
    }
}
